src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1131,143 +839,295 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
 
-	List *nodeToSyncMetadata = NIL;
-	WorkerNode *node = NULL;
-	foreach_ptr(node, nodeList)
+	List *allActivatedWorkerList = GetAllActivatedWorkerList(context);
+
+	List *commandList = NIL;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, allActivatedWorkerList)
 	{
 		/*
-		 * First, locally mark the node is active, if everything goes well,
-		 * we are going to sync this information to all the metadata nodes.
+		 * We need to prevent self deadlock when we access pg_dist_node using separate
+		 * connection to localhost. To achieve this, we check if the caller session's
+		 * pid holds the Exclusive lock on pg_dist_node. After ensuring that (we are
+		 * called from parent session which holds the Exclusive lock), we can safely
+		 * update node metadata by acquiring the relaxed lock.
 		 */
-		WorkerNode *workerNode =
-			FindWorkerNodeAnyCluster(node->workerName, node->workerPort);
-		if (workerNode == NULL)
-		{
-			ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", node->workerName,
-								   node->workerPort)));
-		}
+		StringInfo metadatasyncCommand = makeStringInfo();
+		appendStringInfo(metadatasyncCommand, CITUS_MARK_NODE_NOT_SYNCED,
+						 parentSessionPid, workerNode->nodeId);
+		commandList = lappend(commandList, metadatasyncCommand->data);
+	}
 
-		/* both nodes should be the same */
-		Assert(workerNode->nodeId == node->nodeId);
+	SendCommandListToWorkerOutsideTransactionWithConnection(connection, commandList);
+	CloseConnection(connection);
+}
 
-		/*
-		 * Delete existing reference and replicated table placements on the
-		 * given groupId if the group has been disabled earlier (e.g., isActive
-		 * set to false).
-		 *
-		 * Sync the metadata changes to all existing metadata nodes irrespective
-		 * of the current nodes' metadata sync state. We expect all nodes up
-		 * and running when another node is activated.
-		 */
-		if (!workerNode->isActive && NodeIsPrimary(workerNode))
+
+/*
+ * SetNodeMetadata sets isactive, metadatasynced and hasmetadata flags locally
+ * and, if required, remotely.
+ */
+static void
+SetNodeMetadata(MetadataSyncContext *context, bool localOnly)
+{
+	List *allActivatedWorkerList = GetAllActivatedWorkerList(context);
+
+	/* do not execute local transaction if we collect commands */
+	if (!MetadataSyncCollectsCommands(context))
+	{
+		List *updatedActivatedNodeList = NIL;
+
+		WorkerNode *node = NULL;
+		foreach_ptr(node, allActivatedWorkerList)
 		{
-			bool localOnly = false;
-			DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
-															localOnly);
+			if (NodeIsPrimary(node))
+			{
+				node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_isactive,
+												BoolGetDatum(true));
+				node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_metadatasynced,
+												BoolGetDatum(true));
+				node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_hasmetadata,
+												BoolGetDatum(true));
+			}
+			else if (NodeIsSecondary(node))
```

Comment:
We definitely shouldn't do something like this. Why we need anything related to secondary nodes? We never activate secondary nodes, all we should do is to have a single row in pg_dist_node about the secondaries.

Reply:
I agree secondary node list adds confusion to API. I removed those but we need to address why `citus_add_node` activates secondary nodes in another PR.